### ISSUE TITLE
RFC: Add Model Schemas to OpenApi Docs

### ISSUE THREAD

**Issue by chuckstock** [jlalmes/trpc-openapi#157]
First, I'm loving the project so far and I really appreciate all the work you've done already.

I honestly don't mind the output requirement on a query/mutation, however, it would be nice to be able to define Model Schema's and reference those for the outputs inside the docs.  (ex. Pet Model from [Redoc](https://redocly.github.io/redoc/#tag/pet_model)).

**Comment by jlalmes** [comment]
Hi @chuckstock. I have added this feature as a `Maybe` in our `v1.0.0` roadmap (https://github.com/jlalmes/trpc-openapi/issues/91). Any idea how you would like this API to look?

**Comment by jlalmes** [comment]
Hi @chuckstock. I have added this feature as a "**_Maybe_**" in our `v1.0.0` roadmap (https://github.com/jlalmes/trpc-openapi/issues/91). Any ideas how you would like this API to look/behave?

Initial thoughts 👇

```ts
// models.ts
export const User = z.object({
  id: z.string().uuid(),
  name: z.string(),
});

// router.ts
import { User } from './model';
export const appRouter = trpc.router().query('getUsers', {
  meta: { openapi: { ... } },
  input: ...,
  output: z.array(User),
  resolve: ...
});

// openapi.json.ts
import { User } from './model';
import { appRouter } from './router';

const openapi = generateOpenApiDocument(appRouter, {
  ...,
  models: [User]
});
```

cc @StefanTerdell.

**Comment by chuckstock** [comment]
Yeah this is almost exactly what I was thinking and was potentially expecting to find when reviewing the documentation. I think that makes a lot of sense without adding really any extra boilerplate or overhead.

Another package allows you to extend `ZodObjects` and turn them into openapi components, here is an example from their npm packaage:
```javascript
import { extendApi, generateSchema } from '@anatine/zod-openapi';
const aZodExtendedSchema = extendApi(
      z.object({
        uid: extendApi(z.string().nonempty(), {
          title: 'Unique ID',
          description: 'A UUID generated by the server',
        }),
        firstName: z.string().min(2),
        lastName: z.string().optional(),
        email: z.string().email(),
        phoneNumber: extendApi(z.string().min(10), {
          description: 'US Phone numbers only',
          example: '555-555-5555',
        }),
      }),
      {
        title: 'User',
        description: 'A user schema',
      }
    );
const myOpenApiSchema = generateSchema(aZodExtendedSchema);
```

This is another suggestion, although I think I prefer your original suggestion for the API.

**Comment by OscBacon** [comment]
Hi, any updates on this?

**Comment by rehanvdm** [comment]
I am also very interested in this, its the only open issue on the previous roadmap. Can this still be looked at for a future release?

**Comment by StefanTerdell** [comment]
I'm a bit late on the ball here haha. GHs UI sure is one of the UIs of all time

Dont know much about open api but I believe it moves the reused schemas out of what would be the idiomatic JSON schema position ("#/definitions" or "#/$defs"). But that's what the definitionPath option is for. It still presumes the root path to be the schema root though, but you can get around that with basePath.

Here's a basic example:

```typescript
const baz = z.object({})

zodToJsonSchema(z.object({ x: baz }), {
  basePath: ["foo"],
  definitionPath: "bar",
  definitions: { baz },
})
```
\>
```js
{
  type: 'object',
  properties: { x: { '$ref': 'foo/bar/baz' } },
  required: [ 'x' ],
  additionalProperties: false,
  bar: {
    baz: { type: 'object', properties: {}, additionalProperties: false }
  }
}
```

You should be able to just move out the definitions object ("bar" in the example) from the resulting schema to put wherever you want in the open api doc as long as you can know the path beforehand and point at it with definitionsPath

**Comment by devidw** [comment]
I needed support for this also, published a quick implementation: https://www.npmjs.com/package/@devidw/trpc-openapi

Source-Code: https://github.com/devidw/trpc-openapi

All you need to do is pass a object with your model schemas when you generate the document:

```ts
import { generateOpenApiDocument } from "trpc-openapi";
import { z } from "zod";

const SomeSchema = z.object({
  a: z.literal(69),
});

generateOpenApiDocument(appRouter, {
  // ...
  defs: {
    SomeSchema,
  },
});
```
